io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -808,6 +811,18 @@ schedule thread@Thread{
           simstate' = simstate { clocks = Map.insert clockid' clockoff clocks }
       schedule thread' simstate'
 
+    -- we treat negative timers as cancelled ones; for the record we put
+    -- `EventTimerCreated` and `EventTimerCancelled` in the trace; This differs
+    -- from `GHC.Event` behaviour.
+    NewTimeout d k | d < 0 -> do
+      let t       = NegativeTimeout nextTmid
+          expiry  = d `addTime` time
+          thread' = thread { threadControl = ThreadControl (k t) ctl }
+      trace <- schedule thread' simstate { nextTmid = succ nextTmid }
+      return (Trace time tid tlbl (EventTimerCancelled nextTmid) $
+              Trace time tid tlbl (EventTimerCreated nextTmid nextVid expiry) $
+              trace)
```

Comment:
Events are in the wrong order surely? Should say we create something before we say we cancel it :-)

```suggestion
      return (Trace time tid tlbl (EventTimerCreated nextTmid nextVid expiry) $
              Trace time tid tlbl (EventTimerCancelled nextTmid) $
              trace)
```

Reply:
Ha ha, that's not even possible in quantum word :) - though I am not really sure.